projects/packages/plugin-deactivation/src/dialog-template.php

```diff
@@ -0,0 +1,40 @@
+<?php
+/**
+ * Reference template file for deactivation handler dialog.
+ *
+ * This file is not used directly. Copy this file to your plugin and modify it to suit your needs.
+ *
+ * @package automattic/jetpack-plugin-deactivation
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly
+}
+?>
+<main>
+	<h1><?php esc_html_e( 'Are you sure you want to deactivate?', 'jetpack-plugin-deactivation' ); ?></h1>
+	<p class="big"><?php esc_html_e( 'Before you go...', 'jetpack-plugin-deactivation' ); ?></p>
+	<p><?php esc_html_e( "We'd really love your feedback on our plugin.", 'jetpack-plugin-deactivation' ); ?></p>
+	<p><?php esc_html_e( "Just temporarily deactivating, or don't fancy giving feedback? No problem.", 'jetpack-plugin-deactivation' ); ?></p>
+</main>
+<footer>
+	<button 
+		type="button"
+		class="jp-plugin-deactivation__button"
+		onclick="dispatchEvent(JetpackPluginDeactivation.events.close)"
```

Comment:
This is a template and I want the package user to have control over how they want to setup the buttons or if they even want to have the same buttons. For example, jetpack deactivation has links to help resources and missing feedback.

Reply:
The same thing effect can be achieved with IDs or class names. You can attach listeners only if your conditions are met.

Reply:
It can be. But, it will lock the plugin consumer into having specific buttons. For example, down the road, you want to sever the jetpack connection before deactivating. Having access to those events will let you first run the deactivation and then dispatch the events to control the dialog.


Why don't you like the little bit of inline js though?

Reply:
> For example, down the road, you want to sever the jetpack connection before deactivating. 

In that case, I might have a script that deals with it and I can call `JetpackPluginDeactivation.events.close` from the same script as well, and if the js selector isn't coupled to style, then it's all nice and clean.


> Why don't you like the little bit of inline js though?

* It adds unnecessary complexity to the code.
* It ties down method names to HTML
  * Long-term maintainability is reduced. For example, if a method name is changed or updated, you'll have to think about all the template files that consumed the "previous behavior" as well. This may lead to a shotgun surgery to make a simple change.
* Looking at the JavaScript file, the flow isn't immediately clear, - you have to know that the events are going to be triggered from outside the script
* It forces the consumer of the package to be responsible for dispatching the events, which makes it more difficult to consume.

Reply:
Having events has the advantage of containing the reaction to a specific plugin.
- If it was a method that needed to be called, then they would also have to pass the plugin slug as a parameter.
- If we attached the events to specific buttons, then we don't give consumers control of the dialog.

Perhaps we can improve the documentation to better explain the flow around events? 

I think both you and Mark(we discussed this on a 1on1 today) have similar concerns about exposing the button functionality even when unnecessary. Mark suggested that we can hide the button-related functionality behind an overridable method. And, I agree. 

However, before changing directions I think it would be nice to get reviews from outside the team to understand more accurate potential use cases.

Reply:
I'm not arguing against events. I'm arguing that the listeners should be moved to the class.

> If it was a method that needed to be called, then they would also have to pass the plugin slug as a parameter.

I'm not seeing the problem here. Can't you make it automatically set itself up?

> If we attached the events to specific buttons, then we don't give consumers control of the dialog.

Yes, we do. You control events by moving HTML element IDs around. If you need more control, you can resort to whatever method you prefer for the consumer (including using the current approach). What I'm arguing is that you shouldn't be forced or expected to handle the events when consuming the package.

> However, before changing directions I think it would be nice to get reviews from outside the team to understand more accurate potential use cases.

This isn't about changing a direction. I'm talking strictly about moving event attachments from one place to another.

Reply:
The listeners are inside the class. What am I missing? https://github.com/Automattic/jetpack/blob/fa13547dacaeb5cffb02ebbfa8022eeb984408e4/projects/packages/plugin-deactivation/src/assets/js/deactivation.ts#L46-L50

Reply:
> However, before changing directions I think it would be nice to get reviews from outside the team to understand more accurate potential use cases.

I wouldn't hold up for outside reviewers for too long. Most of Jetpack is on a meetup this week. :) 

> The listeners are inside the class. What am I missing?

I thought the same thing, but I think @pyronaur means the `onclick=` listeners. Listening for the click.

Reply:
You don't need to roll your own event system if you expose open/close methods and either make them overridable or use `wp.hooks` (or both). Then you can set up sensible defaults by searching for certain ID selectors (`jdd-open`, `jdd-close`, `jdd-toggle`, `jdd-etc` ).

Those selectors are easy to move or omit. If additional customization is needed, that can be added via a config object or overridable methods or hooks, or all of the above. 

What I'm saying is that you can design a component so that it works out of the box without sacrificing any flexibility.

And to reiterate conversations we've had before - I wouldn't implement the customization part just yet and just focus on providing sensible defaults. If additional options are needed down the road, they can be added via the config object.

To clarify, by config object I mean an object that you pass to your `JetpackPluginDeactivation` upon instantiation, like so:


```js
new JetpackPluginDeactivation('plugin-name, {
  onClose: () => {}
  onOpen: () => {},
  attachEvents: true,
});
```

imho, that config object isn't necessary now, but it can easily be added later if needed without requiring external files like templates to be modified. 

Template files shouldn't be concerned with logic and shouldn't carry the knowledge of what global JavaScript libraries are loaded. `deactivation-dialog.php` is a template file.

Reply:
This is now done via a data attribute on buttons that do things. I like this new solution. Marking this as resolved.